1주차 미션 제출

9wan6zae opened the pull request:
# 1주차 미션 제출 (한광재)

## 접근 방법

* [x] 미션 코드를 구현하기 위해서 어떤 생각들을 했는지 정리해보세요.

- 입력으로 문자열 또는 숫자가 들어와도 정상적으로 동작할 수 있는 방법을 생각했습니다.
- 유지보수성을 높이기 위해 문자열 입력과 회전을 담당하는 컴포넌트와 알림창을 띄우는 모달 컴포넌트를 분리해서 개발했습니다.
- `RotateString.vue` 컴포넌트에서 테스트 케이스 작성을 위해 입력을 props로도 받을 수 있게 했습니다.
- modal을 효과적으로 개발하기 위해 Vue 3의 기능인 `teleport`를 활용했습니다.

## 특이사항

* [x] 구현하면서 궁금했던 점이나, 아쉬웠던 부분이 있으면 적어주세요.

- 테스트 코드 작성이 미숙해서 `모달 값 유효성 테스트`에서 코드를 추상적으로 구현하지 않고 구체적인 값과 함수를 다룬 부분이 아쉬웠습니다. 

## 제출 후 확인

* [x] 본인의 구현 결과를 `dev` 브랜치에 구현하고, `dev` 브랜치에서 `submission` 브랜치로 PR을 보내주세요.
* [x] Pull Request 생성 후 Reivewer에 @externship-master 와 @lkaybob 을 추가해주세요.

9wan6zae committed:
Test: 문자열 회전 테스트 코드 추가

9wan6zae committed:
Remove: example.spec.js 파일 삭제

9wan6zae committed:
Add: RotatingString 컴포넌트 Home.vue에 추가

9wan6zae committed:
Feat: 문자열 회전 기능 구현

9wan6zae committed:
Style: Eslint 적용
Semicolon 누락, Comma 누락

9wan6zae committed:
Test: 알림 모달 값 유효성 테스트 코드 추가

9wan6zae committed:
Feat: 모달 컴포넌트 구현

9wan6zae committed:
Feat: 모달 컴포넌트 추가

모달 열기, 닫기 기능
slot을 통해 전달될 값 정의

9wan6zae committed:
Fix: 버튼 문구 수정

RotateString 컴포넌트의 버튼이 제시된 요구사항과 달라 수정

9wan6zae committed:
Style: 적절하지 않는 옵션 수정

justify-content: end에서 flex-end로 수정

lkaybob commented on mission/tests/unit/rotateString.spec.js:
```diff
@@ -0,0 +1,53 @@
+import RotateString from '@/components/RotateString.vue';
+
+describe('RotateString.vue', () => {
+  test('문자열 입력', async () => {
```
* 문자열 입력과 버튼 클릭에 대한 TC를 나눠주세요
* TC 이름을 좀 더 상세하게 적어주세요 (ex. `버튼 클릭마다 문자열을 회전한다`)

9wan6zae replied:
다른 분들이 보아도 이해가 되게 문장으로 잘 풀어보겠습니다!

lkaybob commented on mission/tests/unit/rotateString.spec.js:
```diff
@@ -0,0 +1,53 @@
+    await button.trigger('click');
+    expect(wrapper.text()).toContain('rojectlionP');
+  });
+  test('숫자 입력', async () => {
```
이미 문자열 입력 TC가 있어 필요없는 TC인 것 같습니다.

9wan6zae replied:
저도 작성하면서 필요한지 아닌지에 대해 고민이 있었는데 앞으로 충분한 판단으로 DRY 코드를 작성하겠습니다!

lkaybob commented on mission/tests/unit/rotateString.spec.js:
```diff
@@ -0,0 +1,53 @@
+    const wrapper = shallowMount(RotateString, {
+      props: { propsString },
+    });
+    const button = wrapper.find('button');
```
Arrange, Assert, Act 단계가 구분되도록 newline을 넣어주세요

https://next.vue-test-utils.vuejs.org/guide/essentials/a-crash-course.html#arrange-act-assert

9wan6zae replied:
다른 분이 보아도 이해가 쉽게 되도록 newline을 통한 분리가 미흡했습니다. 첨부해주신 링크를 통해 다시 학습하겠습니다!

lkaybob commented on mission/tests/unit/rotateString.spec.js:
```diff
@@ -0,0 +1,53 @@
+    const wrapper = mount(RotateString, {
+      props: { propsString },
+    });
+    const { vm } = wrapper;
```
vm에 직접 접근해 검증하는 TC는 최대한 지양해주세요.
* AlertModal.vue의 p tag에 `[data-test="modal-message"]`와 같은 attribute를 넣어서 HTML로 보이는 출력을 검증할 수도 있었을 겁니다

https://next.vue-test-utils.vuejs.org/guide/essentials/easy-to-test.html#do-not-test-implementation-details

9wan6zae replied:
이 부분에 대해서 궁금했었는데 참고하겠습니다! 감사합니다!!

9wan6zae replied:
teleport에서 TC 작성하는 방법을 잘 모르겠습니다.. 
https://next.vue-test-utils.vuejs.org/guide/advanced/teleport.html#interacting-with-the-teleported-component
이것을 참고해서 했는데도 잘 되지 않아서 혹시 조언가능하지 궁금합니다.

lkaybob commented on mission/src/components/RotateString.vue:
```diff
@@ -0,0 +1,62 @@
+  components: { AlertModal },
+  name: 'RotateString',
+  props: {
+    propsString: String,
```
* propsString이라고 명명하면 역할을 알 수 없습니다. 무슨 역할인가요? (defaultString? displayString?)
* Home.vue에서는 props를 사용하지 않는다면 props가 필요할까요?

9wan6zae replied:
테스트 시 props로 값을 전달하려고 했는데 첨부해주신 자료를 통해 input에 대한 TC 작성 방법을 다시 공부하겠습니다

lkaybob commented on mission/tests/unit/rotateString.spec.js:
```diff
@@ -0,0 +1,53 @@
+  test('문자열 입력', async () => {
+    const propsString = 'Projectlion';
+    const wrapper = shallowMount(RotateString, {
+      props: { propsString },
```
props를 이용해 테스트를 하는 것은 <input> 태그를 직접적으로 테스트하는 것이 아닙니다.
* 예. 글자를 입력하고 수정을 할 때의 테스트케이스?
* Vue Test Util의 [Crash Course](https://next.vue-test-utils.vuejs.org/guide/essentials/a-crash-course.html#adding-a-new-todo)를 참고하시면 input을 테스트하는 방법을 파악하실 수 있을 겁니다

lkaybob reviewed (commented):
고생하셨습니다. 요구사항들을 잘 반영해주셔서 구현해주셨네요.
* teleport와 composition API를 사용하신 부분이 돋보입니다.
* 테스트 코드 작성에 관한 코멘트들과 몇 가지 마이너한 코멘트들 드립니다.

chojooyoung commented on mission/src/components/RotateString.vue:
```diff
@@ -0,0 +1,62 @@
+  };
+  return { string, rotate };
+};
+
```
상태관리를 하는것이 마치 React에서 HOOK을 사용하는 것 같네요! Vue.js에서도 이런식으로 구현하는 것 배워갑니다! 👏👏

9wan6zae replied:
넵! vue3 composition API이라고 불러요 말씀하신대로 React의 커스텀 hook처럼 만들어서 사용해서 코드를 모듈처럼 사용할 수 있어 재사용성이 좋아집니다!

9wan6zae replied:
저도 잘 부탁드립니다!

chojooyoung commented on mission/src/teleport/Modal.vue:
```diff
@@ -0,0 +1,73 @@
+    return { isShow, open, close };
+  },
+});
+</script>
```
알람창을 모달 컴포넌트로 따로 구현하신 것 좋아보입니다!!

chojooyoung commented on mission/src/components/AlertModal.vue:
```diff
@@ -0,0 +1,39 @@
+      {{ text }}
+    </p>
+    <footer>
+      <button class="close-btn" @click="cancel">닫기</button>
```
함수명, 변수명은 직관적인 것이 좋다고 생각해요 😀 '닫기' 버튼인 만큼, '취소'를 뜻하는 cancel 이 아닌 close를 활용한 함수명을 쓰면 좀 더 직관적일 것 같습니다!😀

9wan6zae replied:
좋은 의견 감사합니다! 반영할게요!

chojooyoung reviewed (commented):
안녕하세요 광재님~ 미션1 코드 잘 보았습니다. Vue.js를 공부 하신게 티가 많이 난 코드 같아요! 보면서 많이 배울수 있었습니다. 남은 기간도 잘 부탁드립니다!👊

Yugyeong-Ji commented:
그룹3의 지유경입니다!

[잘한점]
일단 코드를 보면서 vue에 익숙하신 분 같다는 생각이 들었습니다. 코드를 보면서 배울 점이 많다는 생각이 들었습니다!

Yugyeong-Ji commented on mission/src/teleport/Modal.vue:
```diff
@@ -0,0 +1,73 @@
+  },
+});
+</script>
+
```
알림창 스타일 지정까지 해주신 점이 좋아보였습니다!

MiJungle commented on mission/src/views/Home.vue:
```diff
@@ -2,17 +2,20 @@
 // @ is an alias to /src
 import HelloWorld from '@/components/HelloWorld.vue';
+import RotateString from '@/components/RotateString.vue';
 
```
안녕하세요 한광재님! 유지보수성을 고려하여 문자열 입력, 회전을 담당하는 것과 알림창을 띄우는 부분을 분리한 점이 인상 깊었습니다.

yeanjae0312 commented:
안녕하세요 한광재님! 그룹3 이연재입니다.

teleport같은 vue3속성을 사용한 것이 인상적이었고, 상태관리 등 배울점이 많다고 느꼈습니다. 앞으로도 잘 부탁드립니다:)

yeanjae0312 commented on mission/src/teleport/Modal.vue:
```diff
@@ -0,0 +1,73 @@
+        <slot />
+      </div>
+    </div>
+  </teleport>
```
vue3의 속성인 teleport를 사용한 것이 버전3를 잘 활용하신 것 같아서 좋아 보입니다..! 모르던 속성을 하나 알아가서 기분이 좋네요 :)

SangWoo9734 commented on mission/src/teleport/Modal.vue:
```diff
@@ -0,0 +1,73 @@
+<template>
+  <teleport to='body'>
+    <div class="modal" v-show="isShow" @click="isShow = false">
+      <div class='modal__container'>
+        <slot />
+      </div>
+    </div>
+  </teleport>
+</template>
+
+<script>
+import { defineComponent, ref } from 'vue';
+
+export default defineComponent({
+  setup() {
+    const isShow = ref(false);
+
+    const open = () => {
+      isShow.value = true;
+    };
+
+    const close = () => {
+      isShow.value = false;
+    };
+
+    return { isShow, open, close };
+  },
+});
+</script>
```
모달창을 직접 구현하신 부분이 인상적입니다. 그리고 teleport라는 것을 활용하신 부분이 저도 몰랐던 부분이라 그 부분도 새롭습니다!

9wan6zae replied:
감사합니다! 이번 extership을 통해 vue3의 새로운 기능을 사용하고 싶어서 녹여봤습니다!

9wan6zae committed:
Fix: 모달 닫기 메서드명 수정

코드 전체적으로 봤을 때 cancel보다 close가 적합해 close로 수정

SangWoo9734 commented:
안녕하세요! 그룹 3 박상우라고 합니다.

### 좋은점
- teleport라는 태그를 사용하신 부분이 저에게도 공부가 된 것 같습니다.
- 모달창을 직접 구현한 부분이 다양한 시도를 해보려고 하시는 것 같아 좋았습니다.

다양한 시도를 해보고 싶어하시는 것이 느껴졌습니다! 앞으로 리뷰하면서 많은 공부가 될 것 같습니다!
잘부탁드립니다!

9wan6zae committed:
Fix: 필요없는 TC 제거

문자열 입력에 대한 TC가 이미 있기 때문에 제거

9wan6zae committed:
Fix: TC 수정

- Arrange, Assert, Act 단계로 구분되도록 분리

- 가능한 한 vm 객체를 통해 직접 접근하는 코드 지우기
하지만, teleport의 경우 예제가 별로 없어 어려움을 느낌
-> 확인 후 수정 예정

9wan6zae committed:
Fix: TC에 맞게 input, button에 date-test 추가

9wan6zae committed:
Test: 렌더링 관련 TC 추가

9wan6zae committed:
Fix: TC에 맞게 p 태그에 date-test 추가
